### ISSUE TITLE
Some bugs and issues

### ISSUE THREAD

**Issue by Shafticus** [hankmorgan/UnderworldExporter#20]
Just wanted to say awesome job on the progress.  Here are a few things I found so far:

- The healing fountain on level 1 can be picked up and may have no weight at all.  A quick fix might be to make it really heavy.
- Seems like a glitch with sceptres, but I kept one in a gold coffer with other trinkets.  When I changed levels(or maybe loaded a game), it would duplicate to 256, overburdening me.  Tried getting rid of duplicates, but they would just appear again later.
- After rescuing Murgo, I tried talking to him and got no text.  Unable to exit the conversation and had to kill the program.
- Unable to advance sword.  Oddly enough, other mantras seem to work fine/reliably

Thank you for your hard work on this project.

**Comment by Shafticus** [comment]
One other thing.  I defeated the monster of the mines on level 2, but was unable to claim a reward from Goldthirst.  Not sure if it was because I defeated it before talking to him or if the quest was bugged anyway.  Dialog option for defeating the creature was not present.

**Comment by hankmorgan** [comment]
Thanks for spotting these.  I'll be uploading a new release shortly with some fixes including some of those.

- The mantra error is because I accidentally assigned that mantra to Acrobat. This is now fixed.
- I've stopped the fountains from becoming pickable. This is a strange one because I respect both the weight and can be picked up flags in game data files. The fountain object looks like it has zero weight. I've put in a temp fix on this.
-I was unable to replicate the issue with Goldthirst. You should get the option if you tell him you are in search of adventure. If you have one can you upload a save file. 
-I think I know what is the reason for the issue with the sceptres. Is it a particular sceptre you had this issue with? I only tested with the one on the first level. Basically when wand-like objects move between levels there is a second hidden spell object that needs to travel with them. So I think the 256 you are seeing is the new index of the hidden spell object and the sceptre is suddenly acting as a quantity object of that amount.  Again I'd need to see a save file to look into this some more.
-For Murgo. Again I couldn't replicate. After giving the food to the lizard and then talking to murgo he vanished as expected.

**Comment by Shafticus** [comment]
For Goldthirst, I slew the monster first, then told him I would kill the beast.  When I spoke to him any time after that, the only option was to tell him I was still looking.  I will provide the save file.

For the sceptre, yes, I picked up the one on the first level first.  So I think we are in agreement on that one.

Odd on Murgo.  I gave the lizard food, then the portillus opened, I tried to speak to Murgo, but there was no text.  No way to end the conversation that I could tell. If I can replicate, I will send you a save file.

**Comment by Kilg0reTr0ut** [comment]
I'm having issues with that same section with Murgo. I can't give the Lizard guy enough food. It seems food won't stack in my inventory. Also, when I import this into the vanilla UW installation, there are all kinds of problems.

Here are the saves.

[UW Saves 1-4 Urgo problem.zip](https://github.com/hankmorgan/UnderworldExporter/files/1670779/UW.Saves.1-4.Urgo.problem.zip)

**Comment by Kilg0reTr0ut** [comment]
These are some later saves I'd like you to look at. They are further along in the game. If you find out what the issues are in the previous saves, maybe you can apply the process to these. 

Also, I realized that appraise isn't working. I can't tell the condition of items at all (serviceable, worn etc)
these descriptions don't show up

[UW Saves 1-4 (later saves).zip](https://github.com/hankmorgan/UnderworldExporter/files/1670942/UW.Saves.1-4.later.saves.zip)

**Comment by hankmorgan** [comment]
Shafticus: I think the issue with Murgo going "silent" was to do with merging items together. This was deleting one of the merged objects from the game world but it was not telling my object list that the object did not exist anymore. This was causing the conversation virtual machine to crash out.

Kilgore: The first issue is to do with summoned food items. I'm updating the code to do some sanity checking to the food type objects. That should retroactively fix the food stacking issues you were having and allow you to give them to Setharee.  I'm not sure what specific issues there are in the second set of files but first glance they appear okay.  
There is a later issue involving the item you left with Shak for repairing. I'm looking into it now but he will not give you back the item after you pay him. This does not stop him from fixing the sword of justice later on.

I have a few other issues to look at before I can upload the next update.

**Comment by hankmorgan** [comment]
Appraise is the skill for judging the value of deals. It is not fully implemented. Explicitly using the Look interaction mode will allow you to see the state of weapons and armour.

**Comment by hankmorgan** [comment]
Re Inventory bugs. I've spent enough time making new bugs of my own that I have no time to re-implement Looking Glasses bugs

**Comment by Kilg0reTr0ut** [comment]
Thanks, Hank. I just added my later save files in case the inventory stack fix need to be applied to those. By the sounds of it, your next build will take care of that. With appraise, I don't remember having to explicitly right click to appraise items when looking at them, but I checked and that works fine. Probably just me being a dope.

**Comment by Shafticus** [comment]
Ahh, yes.  I remember some Shak issues as well.  First, he wanted > 160 gold to repair a longsword. Seemed really high, but that system may not be fleshed out yet, or maybe I misremembered how expensive the service was.  When I took the sword back and declined the offer, next time I talked to him, he asked for 0 gold.  Had to give him 1 piece.

**Comment by Shafticus** [comment]
Found a few more game breakers dealing with quests.
- Level 4 After killing Roderick, the wounded knight near him does not recognize it.
- Level 4 - joined the knights, got writ and turned it in.  Got quest for the gold plate.  Tried to return the gold plate and he says that is not it. Returns to a state prior to me joining the order.  I could joing again, but now he wants the writ, which was already turned in.  Not 100% sure, but I think I joined/unjoined several times, sometimes with dialogue options being skipped.

**Comment by Shafticus** [comment]
Can't seem to reproduce the issue, but he consistently tells me that this is not the correct plate.  I assure you it is.  Also if you leave the plate(or probably any item) in the trade window when the conversation ends, it will still be there when you get back.  Should probably pop it back into your inventory or on the ground.

Here is the save with the plate in the main inventory and the quest active for the plate.

[save1.zip](https://github.com/hankmorgan/UnderworldExporter/files/1674582/save1.zip)

**Comment by Shafticus** [comment]
Downloaded newest version.  Here is what I found:

Tried to swap port with a pole(stick?) so I could put it in a different container.  Port disappeared from inventory.  Tried to reload an older save to try again, but when I quaffed the green floating potion from Ironwit's quest on level 2, playing screen went black(although the icon for flying did appear).  It did work correctly when I tried it before making that save just a few minutes prior.

**Comment by Shafticus** [comment]
It seems that exiting the game and starting it up again fixes the issue with the block screen.  Happened twice to me with floating potions.  Tried the Goldthirst quest again and talked to him first, then killed beholder, then talked again, and it worked.  Same with writ quest on level 4, and with killing Roderick.  Was able to get the standard.  Brought back the iron plate, but that does not appear to work.  he says that is not the plate.  My theory on some of the odd glitches is that doing something the code doesn't expect messes something up internally, causing further issues.  Not sure about some of the quests though.  Will have to experiment further to see if I can reliably reproduce these issues.  Going to try Shak again with this new game.  Tried to have him repair Rodericks sword on previous version and when I paid him, I got nothing back.

**Comment by Shafticus** [comment]
Typos all over the place.  Screen was black, not block.  Plate was gold not iron.  Got the plate from the maze behind Roderick on level 4, in the secret room with the gold and silver switches.

**Comment by Shafticus** [comment]
Ventured a bit further.  Here are the issues I can remember:

- Shak will quote time left to repair as "XXC2" minutes where XX is the number of minutes.  Maybe reading an extra byte in?
- Saving game message has escape characters in it.
- Level listed on save description is one less than actual level.  Maybe tracking the level with an array and using the index?
- One of the seers on level 6 had an escape character in the dialogue.  Was female I believe.
- Empty lantern can't be refilled with oil?  Tested this out a little.  Tried to refill when lit, get appropriate message.  Try to refill when not lit and not empty, it at least says you refilled it.  If you let the oil run out and try to refill, you get an item error.  Maybe empty lantern is different from non empty lantern?
- Tried level one magic missile spell.  Just expends mana, doesn't appear to do anything.  Maybe that is not fleshed out yet?
- Freed Murgo fine this time.  When reporting to seer on level 6, Owl does not acknowledge the rescue.
- Died from combat getting overzealous with some goblins.  Fighting continued after death, with one goblin likely killing another by trying to hit me with sling.  I did respawn at the silver sapling though.
- In general, dialogue is a bit wonky and may be what is causing some of the quest issues.  Sometimes, person is waiting for a trigger to continue dialogue, like retrieving an item or doing something.  Sometimes, they skip into the conversation, conversation may end abruptly before I can pick an option, or sometimes dialogue resets to the beginning.  I need to investigate more on this but it may be taking some extra input somehow.
- Disappearing items.  Trying to exchange two items sometimes consumes one.  Like trying to exchange port with a pole did it for me.  Had this happen a few other times as well.  Need to investigate more on this.
- Shak continues to offer misleading costs for reparing items.  Usually quotes 160+ gold, then asks for 30 later.
- Shak will sometimes request money even after item is repaired, then give you nothing in return.

Overall, things are progressing relatively smoothly.  Most of these issues are minor and few are really game breaking.

**Comment by Shafticus** [comment]
Finished Ranthru's quest for a book and was granted more casting power.  Was around 30 or so, now lore is at 200 and casting at "00".

**Comment by Shafticus** [comment]
Forgot about this one.  Ghoul that makes boots only took scales when given scales and 1 string.  String went back into my inventory.

**Comment by Shafticus** [comment]
Anjor is unable to count the samples on level 5, though he knows the true comparison.  Offered 79, he said he needed 80, but told me I offered him over 1000.  Give him 80 and he was fine.

**Comment by Shafticus** [comment]
Level 7, unable to use shard to destroy orb.  Shard just disappears.  Even disappears when you drop it on anything.  Also, the orb can be picked up.

**Comment by Shafticus** [comment]
Final additions for this run through.

- Garamon's bones and tri part key gets duplicated on level change
- Bones in general have charges
- Unable to use Garamon's bones on his grave
- Dreams in general are like the quests where the just reset and start over.  After killing Tyball, was not able to get sequence where he asks for his bones to be buried.
- Was able to open door with tripart key, but unable to throw in talismans successfully.  Is the end sequence triggered by the Garamon dream sequence?
- Lockpick skill was 16, but I was able to lockpick any door in one try.
- Not having room for an item in your main inventory causes item to float in the air.  In some cases, item is not retrievable, like the picture of Tom.

**Comment by Shafticus** [comment]
A few issues I forgot:
- Was not able to rescue princess.  Opened gate, but nothing.  Thought there was a cutscene, but unable to interact.
- Odd inventory issue where if you take an item or container that is already inside a container and make it go into your main inventory(by clicking on the container), the weight of what you picked up was subtracted from your encumbrance, but was not added back after the move.  If I take the item from main inventory and put it back in the container or another container, it updates to the correct value again.

**Comment by Shafticus** [comment]
A few more that popped into my head.
- The medallion on level 7 in the cave system is not retrievable.  I can see it, but can't interact with it.
- More of a question here.  It seemed like the level 8 rings of magic and health regeneration weren't working, but it may have just been really slow.  Are the magical items on level 8 working?

**Comment by hankmorgan** [comment]
Hi Shafticus,

I'm looking through these issues and will report back in a couple of days.  If you have any save games to upload then that would be of help. Thanks.

**Comment by hankmorgan** [comment]
Also make sure you using the latest release. I put some fixes to how Shak handled objects in last upload (V1.04)

**Comment by Shafticus** [comment]
Yes, I started a new game when you released that version.  Here are 4 saves that illustrate a few of the issues.  All medallions appear affected, not just the one in the cave.  Tyball's medallion can't be picked up either.

- One save is the cave medallion
- One save is the level 7 chamber with the princess.  You can check out the medallion, princess, and orb issues there.
- One is the level 8 chamber with the door open.  One of my bags has the talismans in it.
- I think one save is at garamons grave on level 5 filled with hundreds of his bones.

You can probably use your dev powers to teleport to other areas if you need to check on other reported issues.

Once you put in your fixes, let me know and I will start a new game if you want and report a new list under a different thread if that helps.
[saves.zip](https://github.com/hankmorgan/UnderworldExporter/files/1692249/saves.zip)

**Comment by hankmorgan** [comment]
Found some time to working through the list. I'm starting to see additions that I made to support UW2 are causing a few weird issues with UW1.  My responses as follows:

    Shak will quote time left to repair as "XXC2" minutes where XX is the number of minutes. Maybe reading an extra byte in?
-This is an unimplemented feature for formatting strings.

    Saving game message has escape characters in it.
-This is how the game stores the string. Low priority to fix.

    Level listed on save description is one less than actual level. Maybe tracking the level with an array and using the index?
-Yeah. I leave it that way for my own convenience.

    One of the seers on level 6 had an escape character in the dialogue. Was female I believe.
-See above.

    Empty lantern can't be refilled with oil? Tested this out a little. Tried to refill when lit, get appropriate message. Try to refill when not lit and not empty, it at least says you refilled it. If you let the oil run out and try to refill, you get an item error. Maybe empty lantern is different from non empty lantern?
-Tried to recreate this. Could not.

    Tried level one magic missile spell. Just expends mana, doesn't appear to do anything. Maybe that is not fleshed out yet?
-I think this was related to the issue with the level 4 object list  I mentioned in the other support thread. I fixed that and I can't replicate this issue.

    Freed Murgo fine this time. When reporting to seer on level 6, Owl does not acknowledge the rescue.
I think there is a bug with the game quests getting wiped. Again I think this was caused by the level 4 bug.

    Died from combat getting overzealous with some goblins. Fighting continued after death, with one goblin likely killing another by trying to hit me with sling. I did respawn at the silver sapling though.
-AI is dumb as a ditch basically.

    In general, dialogue is a bit wonky and may be what is causing some of the quest issues. Sometimes, person is waiting for a trigger to continue dialogue, like retrieving an item or doing something. Sometimes, they skip into the conversation, conversation may end abruptly before I can pick an option, or sometimes dialogue resets to the beginning. I need to investigate more on this but it may be taking some extra input somehow.
-I've rarely come across this but I think you are on to something regarding input. I'll look into it.

    Disappearing items. Trying to exchange two items sometimes consumes one. Like trying to exchange port with a pole did it for me. Had this happen a few other times as well. Need to investigate more on this.
-There was a bug with the pole's code. Fixed.

    Shak continues to offer misleading costs for reparing items. Usually quotes 160+ gold, then asks for 30 later.
-I think Shak is returning the gold purchase value for the sword. Some quick testing shows he will quote that value but then charge a lower value when I come back to pay. As the conversation is essentially a virtual machine running within the game finding issues with these conversations is time consuming. I'll look it to this in detail in the coming days.

Shak will sometimes request money even after item is repaired, then give you nothing in return.
-I want to say this issue is fixed now due to updates I've made to object trading.

Finished Ranthru's quest for a book and was granted more casting power. Was around 30 or so, now lore is at 200 and casting at "00".
-Fixed. 

Forgot about this one. Ghoul that makes boots only took scales when given scales and 1 string. String went back into my inventory.
-This is fixed now. The conversation was a rare case where more than one item is taken from the player. This was also causing issues with the ghoul taking the wrong items from the player inventory. The fix is a bit of a hack though so I want to test a few more conversations before releasing the update.

Anjor is unable to count the samples on level 5, though he knows the true comparison. Offered 79, he said he needed 80, but told me I offered him over 1000. Give him 80 and he was fine.
-This was a bug in the number value string replacement code. Fixed now.

Level 7, unable to use shard to destroy orb. Shard just disappears. Even disappears when you drop it on anything. Also, the orb can be picked up.
-Couple of things here. 1 The share is not used to destroy the orb. You need to use an orb rock from level 8. Regardless there was a data error with the settings for the shard and the orb rock that made them unusable. 



Garamon's bones and tri part key gets duplicated on level change Bones in general have charges 
-The game was tricking itself to thinking the objects in question were magically enchanted. This is now fixed. I've tested this with the bones and I think the same fix will work for the key. I don't have a save game at hand that is close enough to the keys to test yet.

Unable to use Garamon's bones on his grave
-Fixed

 Dreams in general are like the quests where the just reset and start  over.  After killing Tyball, was not able to get sequence where he asks  for his bones to be buried. 
-I think this might be the quests getting wiped issue again. I'll look into it.

Was able to open door with tripart key, but unable to throw in  talismans successfully.  Is the end sequence triggered by the Garamon  dream sequence? 
-You cannot use the talismans until Garamon is buried and you have answered his questions. This is vanilla game behaviour. 

Lockpick skill was 16, but I was able to lockpick any door in one try.
-Some skill checks are set to always pass at the moment.

 Not having room for an item in your main inventory causes item to  float in the air.  In some cases, item is not retrievable, like the  picture of Tom.
-This is actually by design at the moment. Vanilla behaviour is to put objects you receive into your mouse cursor. I've yet to implement this behaviour so instead I try to put the item into the current container. When this cannot happen I place the object in front of the player.

The picture of tom was affected by the same issue that broke the share and orb rocks. Fixed


    The medallion on level 7 in the cave system is not retrievable. I can see it, but can't interact with it.
-Same issue as above. 

    More of a question here. It seemed like the level 8 rings of magic and health regeneration weren't working, but it may have just been really slow. Are the magical items on level 8 working?
-Those spell variants are not implemented fully yet.

**Comment by Shafticus** [comment]
Wow!  Thanks for looking at and fixing these issues, and also for clarifying some things I didn't understand.  Once you release the next version, I will start a new game and see what crops up.

**Comment by hankmorgan** [comment]
Ok. 
I've identified the issue with wands. It's actually the reason for the data getting screwy on level 4. Wands are special because the store the spell enchantment in a second "spell" object that is stored on the map rather than in the player inventory. So when you move between maps this object is suppose to follow you to allow the wand to work. In my implementation of this 2 things are happening. 
- When I move the object between levels I was creating a Spell trap object rather than a spell object.
- The event used to spawn the spell was firing on save game reloading and because it was looking for a spell to match with instead of the created spell trap it was producing duplicates. So for each time a save is made/loaded an additional 2 objects are created per wand on top of additional objects created when moving through maps. This was rapidly filling up the object space. I've fixed it so both the correct object type is created and then only once per map.

I've also spotted the issue with the quest flags. The data is saving okay but when I reloaded a save the value was not being read in correctly and causing the values to go to zero. This will be fixed in the next release.

I'm attempting a full playthrough of my own today to see what else I can spot. I'll add to this thread as I spot anything significant.

**Comment by hankmorgan** [comment]
Magic enchantments (potions/wands) seem badly wrong to me today. I'm about half way through the full playthrough and I'm getting a few issues previously reported and some that were not found to date.

I've found the issue with Shak thinking he has something to repair. A value in the conversation virtual machine was over-flowing.

**Comment by hankmorgan** [comment]
Ok. I've played through from start to finish. The game is beatable once more and I've cleared a bunch of bugs along the way. I have a couple of critical items to fix and I want to validate the magic enchantments on potions and wand. Once I've done that I hope to put up a new release by the weekend

**Comment by Shafticus** [comment]
Awesome!  I will be out the next couple of weeks, but I will be sure to test it out when I can.  I would probably open up a new issue thread depending on what I find if that is ok.  Thanks for your hard work on this project.

**Comment by hankmorgan** [comment]
Yeah. I've identified the issue with the wands. The initial tests on the fix - picking up a wand, dropping a wand, moving between levels with the wand and saving the game while holding the wand check out. I want to do a few more tests on the different types of wands throughout the game-world first. That should be sorted by tomorrow night and I'd expect to be in a position to patch up your save game/update the release by friday.

**Comment by Kilg0reTr0ut** [comment]
Excellent. Thanks Hank

**Comment by hankmorgan** [comment]
Here is an fixed version of your most recent save file. To fix it I've done the following
-Added back missing quest flags for a few quests  (Murgo, Goldthirst, Rodrick, etc)
-Repaired the golden plate so Dorna will recognise it.
-I had to destroy a few of your wands as they may cause problems for you. Wands you have yet to pick up will be okay going forward.

I'm uploading the newest release now (v1.05).  It should be online in the next hour or so. Please check the readme as I've changed how the sound tracks work in this release. 

[save4.zip](https://github.com/hankmorgan/UnderworldExporter/files/1708498/save4.zip)

**Comment by Kilg0reTr0ut** [comment]
Everything looks good so far. Music tracks all load fine. Biden is recognized as having returned by Dorna  (even though he seems stuck outside the main door) and Dorna acknowledges the plate. I will be resuming my playthrough at twitch.tv/kilg0re_tr0ut at 10EST tomorrow.
